src/code/actions.js

```diff
@@ -63,6 +69,81 @@ export function startSession(uuid) {
   };
 }
 
+export function checkSession(fbConnected) {
+  const timeNow = new Date().getTime();
+  const timeThen = window.sessionStorage.getItem('lastUpdate');
+  const timeDeltaSeconds = (timeNow - timeThen) / 1000;
+  // two hours
+  const sessionExpired = timeDeltaSeconds > 60 * 60 * 2;
+
+  const portalUser = window.sessionStorage.getItem('portalAuth') === "true";
+  if (portalUser) {
+    // if we never set a session timestamp
+    if (!timeThen) return CONNECTION_STATUS.disconnected;
+    // if we're truly not connected for read/write to firebase
+    if (!fbConnected) return CONNECTION_STATUS.disconnected;
+    // if we are connected to firebase but session has expired
+    if (fbConnected && sessionExpired) {
+      return CONNECTION_STATUS.disconnected;
+    } else {
+      // Update session time
+      window.sessionStorage.setItem('lastUpdate', timeNow);
+      return CONNECTION_STATUS.online;
+    }
+  } else {
+    // we're not a portal user
+    if (fbConnected) {
+      // We can read from firebase, but we're not storing progress crystals
+      return CONNECTION_STATUS.offline;
```

Comment:
1. I've not actually tested via the portal yet, I've had some trouble figuring out how to create an activity for the branch and would welcome an assist! 

2. We launch Geniventure with a token from the Portal, never actually a direct connection to the Portal, so once we use that token to auth with Firebase, we continue with just an authenticated connection to Firebase. Monitoring Firebase as best possible in this code should highlight problems, but since we don't have an actual connection to the Portal, if the connection fails we have to relaunch from a window that has a live, logged-in connection to the portal to re-generate the JWT for a new instance of Geniventure window.